src/sagemaker/workflow/utilities.py

```diff
@@ -37,3 +42,105 @@ def list_to_request(entities: Sequence[Union[Entity, StepCollection]]) -> List[R
+        raise ValueError("One of s3 input, athena dataset definition, or redshift dataset definition need to be given.")
+
+    recipe = {"metadata": {"version": 1, "disable_limits": False}, "nodes": []}
+
+    source_node = {
+        "node_id": str(uuid4()),
+        "type": "SOURCE",
+        "inputs": [],
+        "outputs": [
+            {
+                "name": "default",
+                "sampling": {"sampling_method": "sample_by_limit", "limit_rows": 50000},
+            }
+        ],
+    }
+
+    input_definition = None
+    operator = None
+
+    if s3_uri is not None:
+        operator = "sagemaker.s3_source_0.1"
+        input_definition = {
+            "__typename": "S3CreateDatasetDefinitionOutput",
+            "datasetSourceType": "S3",
+            "name": input_name,
+            "description": None,
+            "s3ExecutionContext": {
+                "__typename": "S3ExecutionContext",
+                "s3Uri": s3_uri,
+                "s3ContentType": s3_content_type,
+                "s3HasHeader": s3_has_header,
+            },
+        }
+
+    if input_definition is None and athena_dataset_definition is not None:
+        operator = "sagemaker.athena_source_0.1"
+        input_definition = {
+            "datasetSourceType": "Athena",
+            "name": input_name,
+            "catalogName": athena_dataset_definition.catalog,
+            "databaseName": athena_dataset_definition.database,
+            "queryString": athena_dataset_definition.query_string,
+            "s3OutputLocation": athena_dataset_definition.output_s3_uri,
+            "outputFormat": athena_dataset_definition.output_format,
+        }
+
+    if input_definition is None and redshift_dataset_definition is not None:
+        operator = "sagemaker.redshift_source_0.1"
+        input_definition = {
+            "datasetSourceType": "Redshift",
+            "name": input_name,
+            "clusterIdentifier": redshift_dataset_definition.cluster_id,
+            "database": redshift_dataset_definition.database,
+            "dbUser": redshift_dataset_definition.db_user,
+            "queryString": redshift_dataset_definition.query_string,
+            "unloadIamRole": redshift_dataset_definition.cluster_role_arn,
+            "s3OutputLocation": redshift_dataset_definition.output_s3_uri,
+            "outputFormat": redshift_dataset_definition.output_format,
+        }
+
+    source_node["operator"] = operator
+    source_node["parameters"] = {"dataset_definition": input_definition}
+
+    recipe["nodes"].append(source_node)
+
+    type_infer_and_cast_node = {
+        "node_id": str(uuid4()),
+        "type": "TRANSFORM",
+        "operator": "sagemaker.spark.infer_and_cast_type_0.1",
+        "parameters": {},
+        "inputs": [
+            {"name": "default", "node_id": source_node["node_id"], "output_name": "default"}
+        ],
+        "outputs": [{"name": "default"}],
+    }
+
+    recipe["nodes"].append(type_infer_and_cast_node)
```

Comment:
Not specifically because of this infer type operation, but the new data does not necessarily align with existing feature group. How does the user know what/how to cast? I'm not too comfortable exposing the underlying Wrangler data type in the PySDK though since it's only used internal so far. Ideally I think this reconcile should be driven from the UI.

Reply:
*I'm not too comfortable exposing the underlying Wrangler data type in the PySDK though since it's only used internal so far.*

I think the exported notebook (export to feature store) exposes the data type (schema) already?

*Another gap is the user need to provide types for the all the columns right now in order to use the trained_parameters feature*

users can describe the feature group definition and convert the type according by themselves, and supply as the optional schema as `trained_parameters`. Otherwise, we could provide such converter too.

Reply:
> I think the exported notebook (export to feature store) exposes the data type (schema) already?

Ah right, but It wasn't too obvious. 

> users can describe the feature group definition and convert the type accordingly by themselves, and supply as the optional schema as trained_parameters. Otherwise, we could provide such converter too.

The feature group definition could be just a subset of all the dataframe columns.

Reply:
> The feature group definition could be just a subset of all the dataframe columns.

but in our case, this helper only takes care of 1-1 mapping, right? otherwise, I assume there must be a transformation operation gets involved. If we choose to codify everything user can do via the visual interface, I guess it's a whole different story.